Various compiler refactorings

tbosch opened the pull request:
…filePath.

The goal of this change is to simplify the emitters,
as we will soon create a new one to emit TypeScript nodes directly.

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)



**What is the new behavior?**



**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...


**Other information**:

tbosch committed:
refactor(compiler): make `OutputAst` contain the moduleName, not the filePath (#16832).

The goal of this change is to simplify the emitters,
as we will soon create a new one to emit TypeScript nodes directly.

tbosch commented on packages/compiler/src/aot/static_symbol_resolver.ts:
```diff
@@ -175,6 +176,25 @@ export class StaticSymbolResolver {
     this.importAs.set(sourceSymbol, targetSymbol);
   }
 
+  importExpr(containingFilePath: string, symbol: StaticSymbol, typeParams: o.Type[]|null = null): o.Expression {
```
Remove from here again!

mary-poppins commented:
The angular.io preview for 3ff94421c83716a72baa24a90f0ea645c64d5ac4 is available [here][1].

[1]: https://pr16832-3ff94421c83716a72baa24a90f0ea645c64d5ac4.ngbuilds.io/

mary-poppins commented:
The angular.io preview for cc3dbfafe6f4f1ecdda13b68a4014e0a8a6785f9 is available [here][1].

[1]: https://pr16832-cc3dbfafe6f4f1ecdda13b68a4014e0a8a6785f9.ngbuilds.io/

chuckjaz commented on packages/compiler/src/aot/compiler.ts:
```diff
@@ -86,31 +85,28 @@ export class AotCompiler {
-      stylesCompileResults.externalStylesheets.forEach((compiledStyleSheet) => {
-        generatedFiles.push(this._codgenStyles(srcFileUrl, compiledStyleSheet, fileSuffix));
+      const componentStylesheet = this._styleCompiler.compileComponent(outputCtx, compMeta);
+      compMeta.template !.externalStylesheets.forEach((stylesheetMeta) => {
```
Consider a comment for the non-null assert.

tbosch committed:
refactor(compiler): emit OutputAst and not sources (#16832)

This is in preparation for creating typescript nodes
directly from `OutputAst` nodes.

mary-poppins commented:
The angular.io preview for e72c26b6904cebe3849f74181b9736dba4e6486d is available [here][1].

[1]: https://pr16832-e72c26b6904cebe3849f74181b9736dba4e6486d.ngbuilds.io/

chuckjaz commented on packages/compiler/src/aot/generated_file.ts:
```diff
@@ -6,6 +6,20 @@
+
 export class GeneratedFile {
-  constructor(public srcFileUrl: string, public genFileUrl: string, public source: string) {}
+  constructor(
```
This doesn't capture the constraint that either that one of `source` or `stmts` must be supplied but not both. Maybe this might be better:

```ts
  source: string | null;
  stmts: Statement[] | null;
  constructor(public srcFileUrl: string, public genFileUrl: string: src: string | Statement[]) {
    if (typeof src === 'string) {
      this.source = src;
      this.stmts = null;
    } else {
      this.source = null;
      this.stmts = src;
    }
  }

tbosch replied:
Thanks, good point.

tbosch committed:
refactor(compiler): allow sync AOT compilation (#16832).

AOT compilation can be executed synchronously now,
if the `ReosurceLoader` returns a string directly
(and no `Promise`).

mary-poppins commented:
The angular.io preview for 9ff389cce7519fb6689ac05f1535a4df8f90d90f is available [here][1].

[1]: https://pr16832-9ff389cce7519fb6689ac05f1535a4df8f90d90f.ngbuilds.io/

tbosch committed:
ci: add vicb as owner to compiler-cli (#16832)

tbosch committed:
refactor(compiler): don’t rely on global reflector (#16832)

Using the global reflector made it impossible
to compile multiple programs at the same time.

mary-poppins commented:
The angular.io preview for 4bcf6344ed43d35a453962556aceeded4eaac5f2 is available [here][1].

[1]: https://pr16832-4bcf6344ed43d35a453962556aceeded4eaac5f2.ngbuilds.io/

vicb commented on packages/compiler/src/view_compiler/view_compiler.ts:
```diff
@@ -436,11 +439,16 @@ class ViewBuilder implements TemplateAstVisitor, LocalResolver {
       let valueType: QueryValueType = undefined !;
-      if (tokenReference(match.value) === resolveIdentifier(Identifiers.ElementRef)) {
+      if (tokenReference(match.value) ===
+          this.reflector.resolveExternalReference(Identifiers.ElementRef)) {
```
could we cache the values or relying on the reflector cache is good enough ?

tbosch replied:
I'd rather rely on the cache for now to not change too many things.

vicb commented on packages/compiler/test/directive_lifecycle_spec.ts:
```diff
@@ -6,9 +6,13 @@
-import {LifecycleHooks as Hooks} from '@angular/core/src/metadata/lifecycle_hooks';
+
+function hasLifecycleHook(hook: Hooks, directive: any): boolean {
+  return hasLifecycleHookImnpl(new JitReflector(), hook, directive);
```
typo ? "Impl"

tbosch replied:
Done

vicb reviewed (approved):
3rd commit LGTM- minor nits

mary-poppins commented:
The angular.io preview for 982e995e477839bac57861b7e2fd88ea6466d554 is available [here][1].

[1]: https://pr16832-982e995e477839bac57861b7e2fd88ea6466d554.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 20cde8a83e2d89ef0eda8bb3b7004746ce42e405 is available [here][1].

[1]: https://pr16832-20cde8a83e2d89ef0eda8bb3b7004746ce42e405.ngbuilds.io/

tbosch commented:
For the last 2 commits: see https://github.com/angular/angular/pull/16871

chuckjaz commented:
This was closed due to deleting the `ngc` branch. Retarget to `master` to reopen.

mary-poppins commented:
The angular.io preview for 20cde8a83e2d89ef0eda8bb3b7004746ce42e405 is available [here][1].

[1]: https://pr16832-20cde8a83e2d89ef0eda8bb3b7004746ce42e405.ngbuilds.io/

mary-poppins commented:
The angular.io preview for c0b5a5a6878a9ba09ec56895a44ae1d2a68cf45c is available [here][1].

[1]: https://pr16832-c0b5a5a6878a9ba09ec56895a44ae1d2a68cf45c.ngbuilds.io/
